OpenRA.Game/Settings.cs

```diff
@@ -312,120 +364,113 @@ public class Settings
+				throw new InvalidDataException("Settings modules must define a YamlNode attribute");
+
+			var module = attribute.Shared ? modules.GetOrDefault<T>() :
+				modules.WithInterface<T>().SingleOrDefault(m => m.ModInstance == mod);
+
+			// Lazily load/create the module on first use
+			if (module == null)
 			{
-				var sectionYaml = yamlCacheBuilder.FirstOrDefault(x => x.Key == kv.Key);
-				if (sectionYaml == null)
+				if (objectCreator != null)
+					module = (T)objectCreator.CreateBasic(typeof(T));
+				else
+					module = Activator.CreateInstance<T>();
+
+				var nodeKey = attribute.Key;
+				if (!attribute.Shared)
 				{
-					sectionYaml = new MiniYamlNodeBuilder(kv.Key, new MiniYamlBuilder(""));
-					yamlCacheBuilder.Add(sectionYaml);
+					module.ModInstance = mod;
+					nodeKey = $"{attribute.Key}@{mod}";
 				}
 
-				var defaultValues = Activator.CreateInstance(kv.Value.GetType());
-				var fields = FieldLoader.GetTypeLoadInfo(kv.Value.GetType());
-				foreach (var fli in fields)
+				var node = unloaded.FirstOrDefault(n => n.Key == nodeKey);
+				if (node != null)
 				{
-					var serialized = FieldSaver.FormatValue(kv.Value, fli.Field);
-					var defaultSerialized = FieldSaver.FormatValue(defaultValues, fli.Field);
-
-					// Fields with their default value are not saved in the settings yaml
-					// Make sure that we erase any previously defined custom values
-					if (serialized == defaultSerialized)
-						sectionYaml.Value.Nodes.RemoveAll(n => n.Key == fli.YamlName);
-					else
+					unloaded.Remove(node);
+					var err1 = FieldLoader.UnknownFieldAction;
+					var err2 = FieldLoader.InvalidValueAction;
+					try
+					{
+						FieldLoader.InvalidValueAction = (s, t, f) =>
+						{
+							var ret = t.GetField(f)?.GetValue(module);
+							Console.WriteLine($"FieldLoader: Cannot parse `{s}` into `{f}:{t.Name}`; substituting default `{ret}`");
+							return ret;
+						};
+
+						FieldLoader.Load(module, node.Value);
+
+						// Override with commandline args
+						foreach (var f in typeof(T).GetFields())
+						{
+							var argName = attribute.Key + "." + f.Name;
+							if (args.Contains(argName))
+								FieldLoader.LoadFieldOrProperty(module, f.Name, args.GetValue(argName, ""));
+						}
+					}
+					finally
 					{
-						// Update or add the custom value
-						var fieldYaml = sectionYaml.Value.NodeWithKeyOrDefault(fli.YamlName);
-						if (fieldYaml != null)
-							fieldYaml.Value.Value = serialized;
-						else
-							sectionYaml.Value.Nodes.Add(new MiniYamlNodeBuilder(fli.YamlName, new MiniYamlBuilder(serialized)));
+						FieldLoader.UnknownFieldAction = err1;
+						FieldLoader.InvalidValueAction = err2;
 					}
 				}
+
+				modules.Add(module);
 			}
 
-			var keysYaml = yamlCacheBuilder.FirstOrDefault(x => x.Key == "Keys");
-			if (keysYaml == null)
+			return module;
+		}
+
+		public void Save(bool commitModules = true)
+		{
+			var output = unloaded.ToList().ConvertAll(n => new MiniYamlNodeBuilder(n));
```

Comment:
```suggestion
			var output = unloaded.ConvertAll(n => new MiniYamlNodeBuilder(n));
```

this will already create a new list with ConvertAll